target/linux/generic/pending-5.4/472-0-mtd-spi-nor-attempt-4K-erase.patch

```diff
@@ -0,0 +1,159 @@
+ 	struct spi_nor_erase_command *cmd, *next;
+ 	int ret;
++	u8 previous_erase_opcode;
+ 
+ 	ret = spi_nor_init_erase_cmd_list(nor, &erase_list, addr, len);
+ 	if (ret)
+ 		return ret;
+ 
++	if (spi_nor_has_uniform_erase(nor))
++		previous_erase_opcode = nor->erase_opcode;
+ 	list_for_each_entry_safe(cmd, next, &erase_list, list) {
+ 		nor->erase_opcode = cmd->opcode;
+ 		while (cmd->count) {
+@@ -1225,10 +1228,14 @@ static int spi_nor_erase_multi_sectors(s
+ 		kfree(cmd);
+ 	}
+ 
++	if (spi_nor_has_uniform_erase(nor))
++		nor->erase_opcode = previous_erase_opcode;
+ 	return 0;
+ 
+ destroy_erase_cmd_list:
+ 	spi_nor_destroy_erase_cmd_list(&erase_list);
++	if (spi_nor_has_uniform_erase(nor))
++		nor->erase_opcode = previous_erase_opcode;
+ 	return ret;
+ }
+ 
+@@ -1239,17 +1246,37 @@ destroy_erase_cmd_list:
+ static int spi_nor_erase(struct mtd_info *mtd, struct erase_info *instr)
+ {
+ 	struct spi_nor *nor = mtd_to_spi_nor(mtd);
++	struct spi_nor_erase_map *map = &nor->params.erase_map;
+ 	u32 addr, len;
+ 	uint32_t rem;
+ 	int ret;
++	uint32_t rem_minimum;
++	u8 i;
+ 
+ 	dev_dbg(nor->dev, "at 0x%llx, len %lld\n", (long long)instr->addr,
+ 			(long long)instr->len);
+ 
+-	if (spi_nor_has_uniform_erase(nor)) {
++	if (spi_nor_has_uniform_erase(nor))
+ 		div_u64_rem(instr->len, mtd->erasesize, &rem);
+-		if (rem)
+-			return -EINVAL;
++
++	/* Find the smallest set bit of uniform_erase_type, and
++	 * find out the erase size it uses.
++	 * Test the erase length is cleanly divisible by it.
++	 */
++	if (spi_nor_has_uniform_erase(nor) && rem) {
++		for (i = 0; i < SNOR_ERASE_TYPE_MAX; i++) {
++			if (map->erase_type[i].size == 0)
++				continue;
++			if (map->uniform_erase_type & BIT(i)) {
++				div_u64_rem(instr->len,
++					    map->erase_type[i].size,
++					    &rem_minimum);
++				if (rem_minimum)
++					return -EINVAL;
++				break;
++			} else if (i == SNOR_ERASE_TYPE_MAX - 1)
++				return -EINVAL;
++		}
+ 	}
+ 
+ 	addr = instr->addr;
+@@ -1260,7 +1287,9 @@ static int spi_nor_erase(struct mtd_info
+ 		return ret;
+ 
+ 	/* whole-chip erase? */
+-	if (len == mtd->size && !(nor->flags & SNOR_F_NO_OP_CHIP_ERASE)) {
++	if (len == mtd->size &&
++	    !(nor->flags & SNOR_F_NO_OP_CHIP_ERASE) &&
++	    !rem) {
+ 		unsigned long timeout;
+ 
+ 		write_enable(nor);
+@@ -1289,7 +1318,7 @@ static int spi_nor_erase(struct mtd_info
+ 	 */
+ 
+ 	/* "sector"-at-a-time erase */
+-	} else if (spi_nor_has_uniform_erase(nor)) {
++	} else if (spi_nor_has_uniform_erase(nor) && !rem) {
+ 		while (len) {
+ 			write_enable(nor);
+ 
+@@ -4258,9 +4287,10 @@ spi_nor_select_uniform_erase(struct spi_
+ 	if (!erase)
+ 		return NULL;
+ 
+-	/* Disable all other Sector Erase commands. */
+-	map->uniform_erase_type &= ~SNOR_ERASE_TYPE_MASK;
+-	map->uniform_erase_type |= BIT(erase - map->erase_type);
++	/* Disable all other Sector Erase commands.
++	 * map->uniform_erase_type &= ~SNOR_ERASE_TYPE_MASK;
++	 * map->uniform_erase_type |= BIT(erase - map->erase_type);
```

Comment:
I have not looked into if anything else (the sector-at-a-time part) uses `map->uniform_erase_type` for uniform erase region spi-nor.